src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1362,11 +1357,42 @@ private SslClientAuthenticationOptions GetSslOptionsForRequest(HttpRequestMessag
             return _sslOptionsHttp11!;
         }
 
-        private HttpConnection ConstructHttp11Connection(Stream stream, TransportContext? transportContext)
+        private async ValueTask<Stream> ApplyPlaintextFilter(Stream stream, Version httpVersion, HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (Settings._plaintextStreamFilter is null)
+            {
+                return stream;
+            }
+
+            try
+            {
+                return await Settings._plaintextStreamFilter(new SocketsHttpPlaintextStreamFilterContext(stream, httpVersion, request), cancellationToken).ConfigureAwait(false);
```

Comment:
Should we throw if the filter erroneously returns null?  Otherwise presumably the call site is going to null ref.

Reply:
Possibly... the return type is ValueTask<Stream> (not nullable), so in normal cases it shouldn't be null. What's the guidance here in general?

Reply:
It shouldn't be, in that anyone building with nullable reference types enabled should get a warning if they may be returning null.  But NRT isn't perfect, such that there are ways of not getting a warning even if null is returned, such a warning can be explicitly suppressed, or someone may not even have nullable reference types enabled.  This code path is already so expensive, an extra null check on our part isn't going to add any measurable cost, and if it helps to avoid a difficult to diagnose failure at least once, it's probably worth having the check.  If we add it here, we probably want it for the connect callback as well.  My $.02; I don't feel super strongly either way.

Reply:
Seems reasonable.